### PR TITLE
Avoid truncation when truncating means longer output

### LINKED ISSUE #6267: Avoid truncation when it makes the output longer
Pytest version 3.10.1, also seen on Ubuntu 19.04:

With `-v`, when the output of a failure is close to the truncation limit, verbosity is *increased*, and the extra words are not useful. In short, this:

```
$ py.test-3 -v | wc
     60     269    3287
$ py.test-3 -vv | wc
     59     272    3292
```
In addition the  truncation message is misleading in the number of truncated lines. With `-v`:

```
E       assert b'<!DOCTYPE H...cation.</p>\n' == b'{}'
E         At index 0 diff: 60 != 123
E         Left contains more items, first extra item: 68
E         Full diff:
E         + b'{}'
E         - (b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Inter'
E         -  b'nal Server Error</title>\n<h1>Internal Server Error</h1>\n<p>The server en'
E         -  b'countered an internal error and was unable to complete your request.  Either'...
E         
E         ...Full output truncated (2 lines hidden), use '-vv' to show
```

With `-vv`:

```
E       assert b'<!DOCTYPE H...cation.</p>\n' == b'{}'
E         At index 0 diff: 60 != 123
E         Left contains more items, first extra item: 68
E         Full diff:
E         + b'{}'
E         - (b'<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">\n<title>500 Inter'
E         -  b'nal Server Error</title>\n<h1>Internal Server Error</h1>\n<p>The server en'
E         -  b'countered an internal error and was unable to complete your request.  Either'
E         -  b' the server is overloaded or there is an error in the application.</p>\n')
```

That looks *one* extra line to me, not two. Possibly lines are counted by splitting on `\n`, and the message has a terminating `\n`, technically leaving an empty line at the end -- but this empty line is not shown in the `-vv` output.

it would be better for pytest to print the one line of error, rather than the two lines making an inaccurate claim about the truncation.

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
The full output of a test is no longer truncated if the truncation message would be longer than
the hidden text. The line number shown has also been fixed.

I also used f-string for faster string formatting in a previous commit, I don't think this warrant another MR though.

Closes #6267

### CODE REVIEW

File: src/_pytest/assertion/truncate.py
Comment (nicoddemus) [inline src/_pytest/assertion/truncate.py:75]: Shouldn't you use `truncated_explanation` here?

```suggestion
    if len("".join(truncated_explanation)) > tolerable_max_chars:
```

File: testing/test_assertion.py
Comment (nicoddemus) [inline testing/test_assertion.py:1221]: Minor, but I think this is easier to digest:

```suggestion
        line = "a" * 10
        expl = [line, line]
        result = truncate._truncate_explanation(expl, max_lines=10, max_chars=10)
        assert result == [line, line]
```

File: testing/test_assertion.py
Comment (nicoddemus) [inline testing/test_assertion.py:1228]: Same as before:

```suggestion
        line = "a" * 10
        expl = [line, line]
        result = truncate._truncate_explanation(expl, max_lines=1, max_chars=100)
        assert result == [line, line]
```

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Hi @Pierre-Sassoulas, sorry for the delay and thanks for the ping.

LGTM, left a few small comments, please take a look. :+1:

File: src/_pytest/assertion/truncate.py
Comment (nicoddemus) [reply src/_pytest/assertion/truncate.py:75]: If so, probably best to update the comment above:

```
# We reevaluate the need to truncate chars following removal of some lines
```

File: testing/test_assertion.py
Comment (Pierre-Sassoulas) [reply testing/test_assertion.py:1228]: Do you want me to refactor the existing tests like this in another pull request first ?

File: src/_pytest/assertion/truncate.py
Comment (Pierre-Sassoulas) [reply src/_pytest/assertion/truncate.py:54]: I think the leading ``\`` in ``\\n`` count as one but is only one char at the end, the other additional character is the ``1`` and would be counted in ``int(math.log10(len(input_lines) - max_lines))`` if we calculate the exact value or in +2 for "99" respectively.

File: testing/test_assertion.py
Comment (nicoddemus) [reply testing/test_assertion.py:1228]: Did not realize other tests were following the same pattern.

I think it is OK to change the new tests now, and possibly open a new PR refactoring the others. :+1:

File: src/_pytest/assertion/truncate.py
Comment (Pierre-Sassoulas) [inline src/_pytest/assertion/truncate.py:74]:
```suggestion
    # We reevaluate the need to truncate chars following removal of some lines
```

File: changelog/6267.feature.rst
Comment (Pierre-Sassoulas) [reply changelog/6267.feature.rst]: Done in [e7a48a2](https://github.com/pytest-dev/pytest/pull/10446/commits/e7a48a2235885192944c2a1f93be91ca7f96621e)

File: src/_pytest/assertion/truncate.py
Comment (Pierre-Sassoulas) [reply src/_pytest/assertion/truncate.py]: Nice catch ! This led to further improvement in [4d2c40f](https://github.com/pytest-dev/pytest/pull/10446/commits/4d2c40fa670062c90b1a0503871caa87622c7620)

**Comment by Pierre-Sassoulas** [comment]
I can cleanup the history and rebase if you want. I'm not sure if the current fail from the CI is due to my changes, so maybe rebasing would help ?
 ```
 `tox.tox_env.errors.Fail: pass_env values cannot contain whitespace, use comma to have multiple values in a single line, invalid values found 'GH_RELEASE_NOTES_TOKEN GITHUB_REF GITHUB_REPOSITORY
 ```

**Comment by nicoddemus** [comment]
> I can cleanup the history and rebase if you want. I'm not sure if the current fail from the CI is due to my changes, so maybe rebasing would help ?

The failure will be fixed in https://github.com/pytest-dev/pytest/pull/10578, once we merge that don't worry that we can rebase this PR. 👍

File: src/_pytest/assertion/truncate.py
Comment (Pierre-Sassoulas) [inline src/_pytest/assertion/truncate.py:52]:
```suggestion
    # removed but is at least 68 characters:
```

File: src/_pytest/assertion/truncate.py
Comment (Pierre-Sassoulas) [inline src/_pytest/assertion/truncate.py:58]:
```suggestion
    # + int(math.log10(len(input_lines) - max_lines)) (number of hidden line, at least 1)
```

**Comment by nicoddemus** [comment]
@Pierre-Sassoulas `main` has been fixed in #10578, could you please rebase or enable maintainers to push to your fork? I did rebase but could not `push --force` to your fork. 👍

**Comment by Pierre-Sassoulas** [comment]
Strange, "allow edit by maintainer" is checked. I rebased, I'm going to check the macos fail later today :)

**Comment by nicoddemus** [comment]
I triggered the job again, and now it passed, so it seems like it was a flaky error, so do not worry.

**Pull request merged**
